src/smime.cpp

```diff
@@ -0,0 +1,355 @@
+#include "smime.h"
+
+#include "mimemultipart.h"
+#include "mimepart_p.h"
+
+#include <openssl/bio.h>
+#include <openssl/err.h>
+#include <openssl/pem.h>
+#include <openssl/provider.h>
+
+#include <QBuffer>
+#include <QFile>
+#include <QLoggingCategory>
+
+Q_LOGGING_CATEGORY(SIMPLEMAIL_SMIME, "simplemail.smime", QtInfoMsg)
+
+using namespace SimpleMail;
+
+SMime::SMime(MimeMessage *message)
+    : _privateKey(nullptr)
+    , _certificate(nullptr)
+    , _certificateCA(nullptr)
+    , _recipsReceiver(nullptr)
+    , _input(nullptr)
+{
+    initOpenSSL();
+    _mimeMessage = message;
+    writeMimeMessageBuffer();
+}
+
+SMime::~SMime()
+{
+    BIO_free(_input);
+    EVP_PKEY_free(_privateKey);
+    X509_free(_certificate);
+    sk_X509_pop_free(_certificateCA, X509_free);
+    sk_X509_pop_free(_recipsReceiver, X509_free);
+    _mimeMessage = nullptr;
+}
+
+void SMime::setKeyFile(const QString &filename, const QString &password)
+{
+    _keyfile  = filename;
+    _password = password;
+    loadPKCS12PrivateKey();
+}
+
+void SMime::setPublicKey(const QString &filename)
+{
+    _publicKeyfile = filename;
+    loadPKCS12PublicKey();
+}
+
+bool SMime::sign()
+{
+    bool ret  = false;
+    PKCS7 *p7 = nullptr;
+
+    if (!_certificate || !_privateKey) {
+        qCDebug(SIMPLEMAIL_SMIME) << "no certificate or private key";
+        return ret;
+    }
+
+    wrapMimeMultiPart();
+    setSignedHeader();
+
+    int flags = PKCS7_DETACHED | PKCS7_STREAM | PKCS7_BINARY;
+
+    BIO *out = NULL;
```

Comment:
null, raw pointers, would be nicer if it was RAII even if with just std::unique_ptr with custom deleter